### ISSUE TITLE
"Product is shippable" should be in Shipping settings, not Product Information

### ISSUE THREAD

**Issue by bugfolder** [backdrop-contrib/ubercart#421]
Currently, the "Product is shippable" checkbox appears on the product edit page in the vertical tab "Product information." But since this setting has to do with shipping, and there is a "Shipping settings" vertical tab, this checkbox ought to be the first item in that tab group.

**Comment by bugfolder** [comment]
A small complication is that "Product is shippable" is in the `uc_product_form()`, while  the "Shipping settings" fieldset is added in `uc_quote_form_alter()`, so if `uc_quote` isn't enabled, there'd be no place to put it.

I would expect that most stores would have `uc_quote` enabled, so the 80% use case would be to put "Product is shippable" there. But perhaps to accommodate the other 20%, if `uc_quote` isn't enabled, then `uc_product` could create its own small "Shipping settings" vertical tab containing just that checkbox.

**Comment by argiepiano** [comment]
I'm not sure taking out "Product is shippable" out of the Product information is a good idea. This setting, as you said, is provided by the uc_product module, and is stored in the uc_products and the uc_order_products tables, so it's fully integrated into the Product workflow. To me this is really a product setting. The Quote module tab has to do with addresses and shipping types...

But if you think this is worth pursuing, a separate tab when Quote is not enabled may work - I would use the same name as the one created by Quote ("Shipping settings")

**Comment by bugfolder** [comment]
I'm putting a version together now, and I think it will work (but we'll see!). I agree that it's associated with products, data-wise, but from a store manager's UI point of view, "shippable" would go with "Shipping settings."

I encountered this when I was converting a product to non-shippable for testing, and so naturally clicked on the "Shipping settings" tab, and of course it wasn't there. Yes, it was visible in tiny type in the vertical tab summary text, so it was easy to find the checkbox, but putting it in the "Shipping settings" tab would take out a tiny bit of friction.

(What would also help is implementing summary text for ALL of UC's various vertical tabs settings. I'm working on that now for this issue, but there's another issue needed to address all the other blank vertical tabs.)

I should have something to review and/or try out shortly.

**Comment by bugfolder** [comment]
OK, there's a few unfinished bits, but I think this is at a stage worth looking at. 

* The major change is that there's now always a "Shipping settings" vertical tab. If module `uc_quote` is not enabled, then it just contains the "Product is shippable" checkbox. And I added vertical tab summary text for this situation.

* If the `uc_quote` module is enabled, then the "Shipping settings" vertical tab contains "Product is shippable" plus the other shipping settings: default shipping type and product pickup address. I also added the default shipping type to the vertical tab summary text.

* Since I was touching the `.js` files, I followed the Backdrop recommended practice and moved them into `/js` subfolders in the modules.

* And I altered the weights of the "Product information" and "Shipping settings" fieldsets so that they appear together at the top of the listing, rather than "Product information" at the top and "Shipping settings" buried farther down in the list.

Two still TBDs:

* The "Default shipping type" vertical tab text doesn't instantaneously update when changing the select, as opposed to the checkbox that does. I'm not sure why.
* I didn't create summary text for the "Product pickup address", because it would have to be a lot of text to reflect all the fields. Maybe it should just say "Default" or "Custom", instead of trying to list the whole address.

So please take a look and let me know what you think.

**Comment by argiepiano** [comment]
Thanks, @bugfolder.

This works fine in my testing. A couple of things:

- For some reason, applying the patch did not move `uc_product.js` to its new location. I assume that when this goes into a new version, and users download it, the whole folder with the ubercart module will replace the old version, assuring that the js file will be in the new location
- There is another instance where you need to be sure to indicate the new location for `uc_product.js`: line 943 of `uc_product.module` still shows the old location (root of `uc_product`).
- Question: when I de-select "Product is shippable" - should the options for "Default product shipping type" and pickup address be hidden, since the product is not shippable? 

I found why changing the "Default product shipping type"  doesn't update the summary. I'll do a comment in your PR.

**Comment by argiepiano** [comment]
Also I agree that creating a summary with the pickup address would be too much for the vertical tab. I think Default or Custom is the way to go.

**Comment by bugfolder** [comment]
Thanks for your notes and testing! Comments:

> I assume that when this goes into a new version, and users download it, the whole folder with the ubercart module will replace the old version

* It should create the new version in the new place, that's for sure; but I seem to recall that there's some scenario where when a file is moved as part of an update, the old version is left in place (this came up when we were moving some test files into their `test` folder). I'll have to look that up; there may need to be an update hook that checks for stray files.

> There is another instance where you need to be sure to indicate the new location for `uc_product.js`

Good catch, fixed.

> Question: when I de-select "Product is shippable" - should the options for "Default product shipping type" and pickup address be hidden, since the product is not shippable?

Yes! `#states` to the rescue. And this is another good UI reason for putting the "shippable" checkbox in the "Shipping settings" vertical tab, so that it's obvious those those fields appear and disappear based on the visible checkbox.

> I found why changing the "Default product shipping type" doesn't update the summary.

Thanks, fixed!

All updated. Still to-do is to check on whether the js files get moved with no strays left behind.

**Comment by bugfolder** [comment]
Oh, incidentally, I changed the wording under the Product Pickup Address. Testing revealed that all of the parts of that address are ignored (not saved) unless the Street Address (first line) is non-empty. That is, if you fill in everything else but leave that line blank, nothing is saved; whereas if you fill in that line and leave anything else blank. it's all saved.

**Comment by bugfolder** [comment]
> Still to-do is to check on whether the js files get moved with no strays left behind.

Yeah, I knew this rang a bell. It happened before in core. https://github.com/backdrop/backdrop-issues/issues/5650 created the situation, https://github.com/backdrop/backdrop-issues/issues/5465 fixed it. In a nutshell: when you move files within a module, people who upgrade by replacing the module get the new directory structure correctly, but people who upgrade via the project installer will be left with copies of old file(s) in their original location(s). These need to be cleaned up via an update hook, like `system_update_1085()`. So I'll have to add this.

**Comment by bugfolder** [comment]
Now added. Tested the update hook with and without the extra js file being present.

**Comment by argiepiano** [comment]
I'm afraid this is not working with the following steps:
1. I started with a clean install and the current dev version of Ubercart. 
2. I manually patched it with this PR 
3. I ran database updates

The file uc_product.js was removed from the root of uc_product by the update hook as expected, but it does not exist in its new location. 

I can't really tell if this problem will go away when installing the new version from the Project Installer. Do you know?

**Comment by argiepiano** [comment]
Other than the above problem, the feature WFM!

**Comment by bugfolder** [comment]
The problem you experienced is a limitation of `git apply`, which doesn't handle moved files. Do this:

* Generate your patch in usual way, by appending `.patch`to the PR's URL;
* Move `423.patch` to Ubercart root;
* In the shell, type `git am 423.patch`.

That does the right thing.

I learned that from https://stackoverflow.com/questions/26111224/how-to-properly-apply-a-patch-with-moved-renamed-files-using-git.

**Comment by bugfolder** [comment]
> That does the right thing.

Although it also puts those commits onto your HEAD (as opposed to `patch`, which just makes the changes without committing anything).

**Comment by argiepiano** [comment]
Thanks for those directions, @bugfolder. That worked - the js file is where it should be.

I have tested again. I noticed that, now, the quote info is hidden when "Product is shippable" is off 👍🏽 . However, the vertical tab still shows the shipping quote summary (even after saving and editing again - see image). 

I think the solution is to edit `Backdrop.behaviors.ucQuoteFieldsetSummaries` and move everything related to the quote summary inside the `if` statement. I'll make a suggestion in the PR.

<img width="293" alt="Screen Shot 2023-01-03 at 4 46 06 PM" src="https://user-images.githubusercontent.com/9938978/210460236-38f028dc-a0f4-499c-bd09-9999e3c81a7f.png">
